File: m4e-client-desktop/src/app/notification/notifyevent.cpp
/**
 * Copyright (c) 2017 by Botorabi. All rights reserved.
 * https://github.com/botorabi/Meet4Eat
 *
 * License: MIT License (MIT), read the LICENSE text in
 *          main directory for more details.
 */


#include "notifyevent.h"
#include <QJsonDocument>
#include <QJsonObject>


namespace m4e
{
namespace notify
{

NotifyEvent::NotifyEvent()
{
}

QString NotifyEvent::toJSON()
{
    return QString( toJSONDocument().toJson() );
}

QJsonDocument NotifyEvent::toJSONDocument()
{
    QJsonObject obj;
    obj.insert( "type", getType() );
    obj.insert( "subject", getSubject() );
    obj.insert( "text", getText() );
    obj.insert( "data",  getData().object() );

    QJsonDocument doc( obj );
    return doc;
}

bool NotifyEvent::fromJSON( const QString& input )
{
    QJsonParseError err;
    QJsonDocument doc = QJsonDocument::fromJson( input.toUtf8(), &err );
    if ( err.error != QJsonParseError::NoError )
        return false;

    return fromJSON( doc );
}

bool NotifyEvent::fromJSON( const QJsonDocument& input )
{
    QJsonObject obj      = input.object();
    QString     type     = obj.value( "type" ).toString( "" );
    QString     subject  = obj.value( "subject" ).toString( "" );
    QString     text     = obj.value( "text" ).toString( "" );
    QJsonObject data     = obj.value( "data" ).toObject();

    setType( type );
    setSubject( subject );
    setText( text );
    setData( QJsonDocument( data ) );

    return true;
}

} // namespace notify
} // namespace m4e

File: src/main/java/net/m4e/app/notification/NotifyEvent.java
/*
 * Copyright (c) 2017-2018 by Botorabi. All rights reserved.
 * https://github.com/botorabi/Meet4Eat
 * 
 * License: MIT License (MIT), read the LICENSE text in
 *          main directory for more details.
 */

package net.m4e.app.notification;

import java.util.Map;

/**
 * Base event used for sending a notification
 * 
 * @author boto
 * Date of creation Oct 12, 2017
 */
public abstract class NotifyEvent {

    /**
     * Sender ID, let it be 0 in the case that the sender is the system.
     */
    private Long senderId = 0L;

    /**
     * Notification type
     */
    private String type = "";

    /**
     * Notification subject
     */
    private String subject = "";

    /**
     * Notification text
     */
    private String text = "";

    /**
     * If needed put some data into the event.
     */
    private Map<String, Object> data;


    public NotifyEvent() {}

    /**
     * Get ID of user sending this notification. It can be 0 meaning that the notification
     * comes from system.
     * 
     * @return Sender ID
     */
    public Long getSenderId() {
        return senderId;
    }

    /**
     * Set ID of user sending this notification.
     * 
     * @param senderId User ID
     */
    public void setSenderId(Long senderId) {
        this.senderId = senderId;
    }

    /**
     * Get the notification type
     * 
     * @return Type
     */
    public String getType() {
        return type;
    }

    /**
     * Set the notification type
     * 
     * @param type Notification type
     */
    public void setType(String type) {
        this.type = type;
    }

    /**
     * Get the notification subject.
     * 
     * @return Subject
     */
    public String getSubject() {
        return subject;
    }

    /**
     * Set the notification subject.
     * 
     * @param subject Notification subject
     */
    public void setSubject(String subject) {
        this.subject = subject;
    }

    /**
     * Get the notification text.
     * 
     * @return Notification text
     */
    public String getText() {
        return text;
    }

    /**
     * Set the notification text.
     * 
     * @param message Notification text
     */
    public void setText(String message) {
        this.text = message;
    }

    /**
     * Get the notification data. It can be null if no data is needed.
     * 
     * @return Data
     */
    public Map<String, Object> getData() {
        return data;
    }

    /**
     * Set the notification data.
     *
     * @param data Data
     */
    public void setData(Map<String, Object> data) {
        this.data = data;
    }
}
